src/middlewared/middlewared/plugins/pool.py

```diff
@@ -4182,6 +4184,148 @@ def __handle_zfs_set_property_error(self, e, properties_definitions):
             raise CallError(pjob.error)
         return data
 
+    @accepts(
+        Str('id', required=True),
+        Ref('query-filters'),
+        Ref('query-options'),
+    )
+    @item_method
+    async def get_userquota(self, id, filters, options):
+        """
+        Returns a list of user quotas on the ZFS dataset `id`. Supports
+        `query-filters` and `query-options`.
+        """
+        quota_list = []
+        userquota = await run(
+            ['zfs', 'userspace', '-H', '-n', '-p', '-o', 'name,used,quota', id],
+            check=False
+        )
+        if userquota.returncode != 0:
+            raise CallError('Failed to get userquota for %s: [%s]',
+                            id, userquota.stderr.decode())
+
+        for quota in userquota.stdout.decode().splitlines():
+            entry = {
+                'uid': None,
+                'username': None,
+                'quota': 0,
+                'used': 0,
+            }
+            m = RE_USERSPACE.match(quota)
+            if m is None:
+                self.logger.debug('Invalid userspace quota: %s', quota)
+                continue
+
+            entry.update(m.groupdict())
+
+            try:
+                entry['username'] = (
+                    await self.middleware.call('user.get_user_obj', {'uid': entry['uid']})
+                )['pw_name']
```

Comment:
Seems we could cache this value in a dict, so we dont do it multiple times for the same uid?